src/components/FilesRealTimeQueries.jsx

```diff
@@ -0,0 +1,130 @@
+
+/**
+ * Normalizes an object representing a CouchDB document
+ *
+ * Ensures existence of `_type`
+ *
+ * @public
+ * @param {CouchDBDocument} couchDBDoc - object representing the document
+ * @returns {CozyClientDocument} full normalized document
+ */
+const normalizeDoc = (couchDBDoc, doctype) => {
+  return {
+    id: couchDBDoc._id,
+    _type: doctype,
+    ...couchDBDoc
+  }
+}
+
+/**
+ * DispatchChange
+ *
+ * @param {CozyClient} client CozyClient instane
+ * @param {Doctype} doctype Doctype of the document to update
+ * @param {CouchDBDocument} couchDBDoc Document to update
+ * @param {Mutation} mutationDefinitionCreator Mutation to apply
+ */
+const dispatchChange = (
+  client,
+  doctype,
+  couchDBDoc,
+  mutationDefinitionCreator
+) => {
+  const data = normalizeDoc(couchDBDoc, doctype)
+  const response = {
+    data
+  }
+
+  const options = {}
+  client.dispatch(
+    receiveMutationResult(
+      client.generateRandomId(),
+      response,
+      options,
+      mutationDefinitionCreator(data)
+    )
+  )
+}
+
+const ensureFileHasPath = async (doc, client) => {
+  if (doc.path) return doc
+
+  const parentQuery = buildFileByIdQuery(doc.dir_id)
+  const parentResult = await client.fetchQueryAndGetFromState(parentQuery)
+
+  return ensureFilePath(doc, parentResult.data)
+}
+
+/**
+ * Component that subscribes to io.cozy.files document changes and keep the
+ * internal store updated. This is a copy of RealTimeQueries from cozy-client
+ * with a tweak to merge the changes with the existing document from the store.
+ * You can have more detail on the problematic we are solving here:
+ * https://github.com/cozy/cozy-client/issues/1412
+ *
+ * @param  {object} options - Options
+ * @param  {Doctype} options.doctype - The doctype to watch
+ * @returns {null} The component does not display anything.
+ */
+const FilesRealTimeQueries = ({
+  doctype = 'io.cozy.files',
+  computeDocBeforeDispatchCreate = ensureFileHasPath,
+  computeDocBeforeDispatchUpdate = ensureFileHasPath,
+  computeDocBeforeDispatchDelete = (doc, client) =>
+    ensureFileHasPath({ ...doc, _deleted: true }, client)
+}) => {
+  const client = useClient()
+
+  useEffect(() => {
+    const realtime = client.plugins.realtime
+
+    if (!realtime) {
+      throw new Error(
+        'You must include the realtime plugin to use RealTimeQueries'
+      )
+    }
+
+    const dispatchCreate = async couchDBDoc => {
+      const doc = await computeDocBeforeDispatchCreate(couchDBDoc, client)
+      dispatchChange(client, doctype, doc, Mutations.createDocument)
+    }
+    const dispatchUpdate = async couchDBDoc => {
+      const doc = await computeDocBeforeDispatchUpdate(couchDBDoc, client)
+      dispatchChange(client, doctype, doc, Mutations.updateDocument)
+    }
+    const dispatchDelete = async couchDBDoc => {
+      const doc = await computeDocBeforeDispatchDelete(couchDBDoc, client)
+      dispatchChange(client, doctype, doc, Mutations.deleteDocument)
+    }
+
```

Comment:
Nice!

nit: I think in order to easily migrate later, you should have pass the method `computeDocBeforeDispatchXXX` to the dispatchChange. But we can start with this one 👍